### PR TITLE
DynamicAtlas: Aggressively load relations in preemptive loading

### PR DESCRIPTION
**Pull request by matthieun**
### Description:

Aggressive loading of relations was triggered every time a relation was asked from a DynamicAtlas. In some use cases, when a DynamicAtlas contains hundreds of relations, the neighboring shards were loaded and parsed hundreds of times each, which was prohibitive.

This includes multiple related enhancements:

- Refactor the expansion logic out of the `DynamicAtlas` class, into `DynamicAtlasExpander`, for clarity.
- Restrict aggressive relation loading to the preemptive load call; effectively limiting its availability to run only when the policy is to defer loading and to not expand indefinitely.
- Better logging of the relation expansion. Now a set of shards being loaded are printed as a WKT GEOMETRYCOLLECTION. Example logs:
```
DynamicAtlas([12-1350-1870]): 
Triggering new shard load for Atlas RELATION 3 with shape 
GEOMETRYCOLLECTION (LINESTRING (-61.3816 15.4852, -61.3949 15.4781), LINESTRING (-61.3949 15.4781, -61.3826 15.4145))
```
and
```
DynamicAtlas([12-1350-1870]): Aggressively exploring relations in shards 
[12-1349-1868, 12-1350-1868, 12-1351-1870, 12-1348-1871, 12-1348-1870, 12-1351-1871, 12-1349-1871, 12-1350-1871, 12-1348-1868, 12-1348-1869, 12-1351-1868, 12-1351-1869] - 
GEOMETRYCOLLECTION (POLYGON ((-61.4355469 15.2841851, -61.4355469 15.3689499, -61.3476562 15.3689499, -61.3476562 15.2841851, -61.4355469 15.2841851)), POLYGON ((-61.2597656 15.2841851, -61.2597656 15.3689499, -61.171875 15.3689499, -61.171875 15.2841851, -61.2597656 15.2841851)), POLYGON ((-61.4355469 15.5383759, -61.4355469 15.6230368, -61.3476562 15.6230368, -61.3476562 15.5383759, -61.4355469 15.5383759)), POLYGON ((-61.2597656 15.3689499, -61.2597656 15.4536802, -61.171875 15.4536802, -61.171875 15.3689499, -61.2597656 15.3689499)), POLYGON ((-61.5234375 15.2841851, -61.5234375 15.3689499, -61.4355469 15.3689499, -61.4355469 15.2841851, -61.5234375 15.2841851)), POLYGON ((-61.2597656 15.4536802, -61.2597656 15.5383759, -61.171875 15.5383759, -61.171875 15.4536802, -61.2597656 15.4536802)), POLYGON ((-61.5234375 15.3689499, -61.5234375 15.4536802, -61.4355469 15.4536802, -61.4355469 15.3689499, -61.5234375 15.3689499)), POLYGON ((-61.2597656 15.5383759, -61.2597656 15.6230368, -61.171875 15.6230368, -61.171875 15.5383759, -61.2597656 15.5383759)), POLYGON ((-61.5234375 15.4536802, -61.5234375 15.5383759, -61.4355469 15.5383759, -61.4355469 15.4536802, -61.5234375 15.4536802)), POLYGON ((-61.5234375 15.5383759, -61.5234375 15.6230368, -61.4355469 15.6230368, -61.4355469 15.5383759, -61.5234375 15.5383759)), POLYGON ((-61.3476562 15.2841851, -61.3476562 15.3689499, -61.2597656 15.3689499, -61.2597656 15.2841851, -61.3476562 15.2841851)), POLYGON ((-61.3476562 15.5383759, -61.3476562 15.6230368, -61.2597656 15.6230368, -61.2597656 15.5383759, -61.3476562 15.5383759)))
```

### Potential Impact:

DynamicAtlas will be way faster when exploring neighboring shards for shared relations.

### Unit Test Approach:

Updated existing unit tests.

### Test Results:

Unit tests pass.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/WktPrintable.java [hunk 1]
```diff
@@ -1,9 +1,22 @@
 package org.openstreetmap.atlas.geography;
 
+import org.openstreetmap.atlas.utilities.collections.StringList;
+
 /**
  * @author matthieun
  */
 public interface WktPrintable
 {
+    static String toWktCollection(final Iterable<? extends WktPrintable> input)
```
Comment (lucaspcram) [inline src/main/java/org/openstreetmap/atlas/geography/WktPrintable.java]: Good stuff.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+import org.openstreetmap.atlas.utilities.collections.StreamIterable;
+import org.openstreetmap.atlas.utilities.collections.StringList;
+import org.openstreetmap.atlas.utilities.time.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author matthieun
+ */
+class DynamicAtlasExpander
```
Comment (lucaspcram) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: So it looks to me like this is a new helper class, but much of the code here is not new and was pulled from `DynamicAtlas`?

Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: Yes, that is correct. `DynamicAtlas` was becoming too big... I understand it makes the review a bit more complicated, so let me add some comments to the lines that are new.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+        // Load all the shards into a multiAtlas
+        buildUnderlyingMultiAtlas();
+        // Record the current list of shards
+        Set<Shard> currentShards = new HashSet<>(this.loadedShards.keySet());
+        // Loop through the entities again to find potential shards to add. This can still happen if
+        // a way intersects the initial shard without shapepoints inside the initial shards, and was
+        // revealed by loading a new neighboring shard. At that point, if that way also intersects a
+        // third shard which was not loaded before, that third shard might become now eligible.
+        browseForPotentialNewShards();
+        browseForPotentialNewShardsFromAggressiveRelations();
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: This is new, it now browses for potential neighboring shards when the policy has the right parameters.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+                        missingShards.stream().map(Shard::getName).collect(Collectors.toList()));
+            }
+
+            // Load all the shards into a multiAtlas
+            buildUnderlyingMultiAtlas();
+            // Record the current list of shards
+            currentShards = new HashSet<>(this.loadedShards.keySet());
+            // Loop through the entities again to find potential shards to add.
+            browseForPotentialNewShards();
+            browseForPotentialNewShardsFromAggressiveRelations();
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: Same here.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+    }
+
+    boolean relationCovered(final Relation relation)
+    {
+        final Set<Long> parentRelationIdentifierTree = new HashSet<>();
+        parentRelationIdentifierTree.add(relation.getIdentifier());
+        return relationCoveredInternal(relation, parentRelationIdentifierTree);
+    }
+
+    private void addNewShardLog(final Shard shard)
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: This concatenates logging in one location to remove duplications.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+            if (!this.loadedShards.containsKey(shard))
+            {
+                this.loadedShards.put(shard, this.atlasFetcher.apply(shard).orElse(null));
+                addNewShardLog(shard);
+            }
+        }
+        final List<Atlas> nonNullAtlasShards = getNonNullAtlasShards();
+        if (!nonNullAtlasShards.isEmpty())
+        {
+            if (shouldBuildUnderlyingMultiAtlasWhenAddingNewShards(initialNonEmptyLoadedShards))
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: Here some complexity has been pushed to this method: `shouldBuildUnderlyingMultiAtlasWhenAddingNewShards `

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+        return this.policy.getInitialShardsBounds().overlaps(area.asPolygon());
+    }
+
+    private void browseForPotentialNewShards()
+    {
+        // Look at regular entities
+        this.dynamicAtlas.entities();
+    }
+
+    private void browseForPotentialNewShardsFromAggressiveRelations()
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: This new method makes sure to browse for new shards and relations only once, and not for every relation considered for expansion.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+    private void newPolyLine(final PolyLine polyLine, final LineItem... source)
+    {
+        if (!loadedShardsfullyGeometricallyEnclosePolyLine(polyLine))
+        {
+            newShapeLog(polyLine, source);
+            addNewShards(this.sharding.shardsIntersecting(polyLine));
+        }
+    }
+
+    private void newShapeLog(final GeometryPrintable geometry, final AtlasEntity... source)
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: This concatenates logging in one location to remove duplications.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 9]
```diff
@@ -0,0 +1,691 @@
...
+            newParentRelationIdentifierTree.add(newIdentifier);
+            if (relationCoversInitialShardBoundsInternal(relation, newParentRelationIdentifierTree))
+            {
+                result = true;
+            }
+        }
+        return result;
+    }
+
+    private boolean shouldBuildUnderlyingMultiAtlasWhenAddingNewShards(
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: The logic here is the same, but the boolean operations have been broken up and commented out for clarity.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java [hunk 10]
```diff
@@ -0,0 +1,691 @@
...
+        // policy is to not extend indefinitely, and to defer loading.
+        if (this.policy.isAggressivelyExploreRelations() && !this.policy.isExtendIndefinitely()
+                && this.policy.isDeferLoading())
+        {
+            // Get all the neighboring shards
+            final Set<Shard> onlyNeighboringShards = new HashSet<>();
+            this.loadedShards.keySet().forEach(
+                    shard -> this.sharding.neighbors(shard).forEach(onlyNeighboringShards::add));
+            onlyNeighboringShards.removeAll(this.loadedShards.keySet());
+            if (logger.isTraceEnabled())
```
Comment (lucaspcram) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java]: This looks like a sonar suggestion :)

**Comment by Huyuntj** [comment]
Looks like a lot of changes here, can't finish reading this year, but have a last question in 2018 :). 
Is the improvement something like, instead of loading same shards several times for different relations, you cache it somewhere and check if new shard existed to avoid loading it multiple time for different relations?

**Comment by matthieun** [comment]
@Huyuntj That is close to it yes! We do not cache the shards, but we look at them once, and process all the relations we need to at once, without re-opening the shards every time. That was a nasty bug.

**Pull request merged**